### PR TITLE
bpo-37355: For nonblocking sockets, call SSL_read in a loop

### PR DESCRIPTION
**Pull request by hashbrowncipher**
Continue looping until data is exhausted, and only then reacquire the GIL. This
makes it possible to perform multi-threaded TLS downloads without saturating
the GIL. On a test workload performing HTTPS download with 32 threads pinned
to 16 cores, this produces a 4x speedup.

                              before     after
    wall clock time (s)  :    29.637     7.116
    user time (s)        :     8.793    12.584
    system time (s)      :   105.118    30.010
    voluntary switches   : 1,653,065   248,484
    speed (MB/s)         :      4733     19712

Also add `PyErr_HasSignals()`, which checks whether a signal has arrived
without holding the GIL.

https://bugs.python.org/issue37355

### CODE REVIEW

File: Lib/test/test_ssl.py
Comment (tiran) [inline Lib/test/test_ssl.py]: ``npn_protocols`` is gone

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c]:
```suggestion
    size_t got = 0;
```

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c]:
```suggestion
                err = _PySSL_errno(retval == 0, self->ssl, retval);
```

File: Modules/signalmodule.c
Comment (tiran) [inline Modules/signalmodule.c]: New public functions must be documented in the C-API section.

@vstinner please take a look

**Review by tiran — CHANGES_REQUESTED** [review]
There are a bunch of C compiler warnings and failing tests. Please take a look.

**Review by vstinner — COMMENTED** [review]
I dislike PyErr_HasSignals() function. You should reuse the existing PyErr_CheckSignals() function. The idea of PyErr_CheckSignals() is that if you hold the GIL for a long time, between two functions calls, you should check for signals. If PyErr_CheckSignals() fails, it raises an exception, you must abandon what you are doing and exit the function (with the exception raised by PyErr_CheckSignals()).

Check for other code using PyErr_CheckSignals() to have examples.

**Comment by hashbrowncipher** [comment]
AFAICT `PyErr_CheckSignals` requires holding the GIL when called. I added `PyErr_HasSignals` to allow checking for signal arrival without the GIL held. If it returns true, then the calling code knows that it must reacquire the GIL and call `PyErr_CheckSignals`.

In the case of this specific code path, it is possible (but somewhat uncommon) for the calls to `SSL_read` to take a noticeable amount of time to fill the entire buffer, even though they are nonblocking. The options I as I saw them were:
1. pay a 4x throughput penalty by reacquiring the GIL and calling `PyErr_CheckSignals` after each 16KB record (status quo)
1. introduce a function like `PyErr_HasSignals` to check for signal arrival without holding the GIL
1. do nothing about signal arrival until the buffer is full.

@vstinner can you see any approaches that I missed? What would you do in this situation?

**Comment by hashbrowncipher** [comment]
I'm aware of the failing `test_ftplib` tests, but I won't be able to debug them until later.

**Comment by njsmith** [comment]
I think you could avoid the `HasSignals` trick by not bothering to check for signals at all during a non-blocking read loop. Checking for signals during blocking operations is important because otherwise they might delay processing the signal for an unbounded amount of time. But for a non-blocking read, it's guaranteed to complete quickly in any case, so it should be enough to check for signals at the end, I think?

**Comment by hashbrowncipher** [comment]
> I think you could avoid the `HasSignals` trick by not bothering to check for signals at all during a non-blocking read loop. Checking for signals during blocking operations is important because otherwise they might delay processing the signal for an unbounded amount of time. But for a non-blocking read, it's guaranteed to complete quickly in any case, so it should be enough to check for signals at the end, I think?

I agree that it's fair to assume non-blocking read operations will complete quickly, but I'm not as confident making that assumption for a loop that could do millions of such operations. Imagine a situation where the application reads an HTTP header that says "Content-Length: 20000000000". It wants to buffer the download in memory, so it calls `sock.read(20000000000)`. If the receiver is CPU-bound, then there will always be data available for processing in the socket buffer, and signals will not be handled until the entire download completes. If it is not CPU bound, then the read operation will return early.

The upshot is that if we don't check for signals until after the end of the loop, the latency of signal processing will depend on whether the download was CPU bound, which seems surprising to me.

**Comment by njsmith** [comment]
Calling `sock.read(20_000_000_000)` is pretty broken to start with – you force the system to allocate a 20 gigabyte buffer, but then python will truncate it to whatever it manages to read before hitting end of data, so 99.99% of the time you allocate 20 gigabytes and then you immediately discard them again. And for normal sockets, that's sufficient protection, because you each call to `.recv` only makes one syscall. I do see your point about if you're CPU-bound though; it's annoying openssl doesn't just do one syscall and then decrypt all of it, but instead forces another syscall for each packet...

File: Modules/signalmodule.c
Comment (hashbrowncipher) [reply Modules/signalmodule.c]: The latest version of the PR removes `PyErr_HasSignals`.

**Comment by hashbrowncipher** [comment]
I have removed `PyErr_HasSignals`. Signals will be buffered until the reading completes.

**Comment by hashbrowncipher** [comment]
The remaining problem is that this change breaks the `test_ftplib` tests. I'd be interested in any bright ideas for how to remedy this, because right now (to me) it looks like this optimization changes existing expectations. 

Right now callers can rely on fact that either SSLSocket.read returns bytes OR it reads a TLS close_notify message, producing an exception; it never does both. With this change, it becomes possible to read application data and also read a TLS close_notify message, which will be processed by OpenSSL. `select/poll/epoll` will all now indicate that the socket has no bytes available (which is true), but OpenSSL knows that no additional bytes will ever arrive on the socket. I validated my theory by patching the test_ftplib code to check for EOF after successful reads. With this change, the tests all passed.

That said, I don't imagine it's acceptable to change behavior like this. I think that to roll out this optimization would require calling code to opt in by passing a flag indicating "Please read eagerly from the socket; I know that I am responsible for explicitly checking for EOF before calling select() on the underlying OS socket." Does anyone have a better idea?

**Comment by Safihre** [comment]
Anyone that can give input? So @hashbrowncipher can continue.

**Comment by dimaqq** [comment]
Should this pull request no longer be a **Draft**?

**Comment by pitrou** [comment]
Does anyone know why the CI is blocked on this PR?

**Comment by tirkarthi** [comment]
It could be due to being a draft PR

https://docs.microsoft.com/en-us/azure/devops/release-notes/2020/sprint-173-update#github-draft-pull-requests-do-not-trigger-pipelines
https://developercommunity.visualstudio.com/t/github-draft-pull-requests-should-trigger-the-pipe/1140412

**Comment by jdevries3133** [comment]
@dimaqq a core developer still hasn't reviewed this to give @hashbrowncipher responses to his questions about how to proceed. @hashbrowncipher, if you put a comment on this post that says, "I have made the requested changes; please review again," the bot will tag this PR as needing core dev review. That might help you get your answers.

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c:2337]:
```suggestion
    size_t readbytes;
```

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c:2341]: Why is it necessary to initialize the variable with ``0``?

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c:2409]: ``got`` and ``count`` are confusing variable names. I suggest that you keep ``count`` as counter for the ``read()`` function call and introduce a new variable for ``SSL_read_ex()`` call.

Is it a good idea to have a potentially infinite loop here that ignores signals and timeouts? This smells like DoS vunlerable to be happen.

```suggestion
            retval = SSL_read_ex(self->ssl, mem + got, len, &readbytes);
            if (retval <= 0) {
                break;
            }

            count += readbytes;
            len -= readbytes;
        } while (nonblocking && len > 0);
```

File: Modules/_ssl.c
Comment (tiran) [inline Modules/_ssl.c:2414]:
```suggestion
        if (count > 0) {
```

File: Lib/test/test_ssl.py
Comment (tiran) [inline Lib/test/test_ssl.py:2303]:
```suggestion
            self.fail("All TLS reads were smaller than 16KB")
```

**Review by tiran — CHANGES_REQUESTED** [review]
Please address my comments and implement the same feature for write().

**Comment by Safihre** [comment]
As @hashbrowncipher seems to have abandoned this (I tried to reach him on Twitter also, but no response), I will rebase and apply the requested changes.

**Comment by dimaqq** [comment]
@Safihre please do your best :)

**Comment by Safihre** [comment]
@tiran/@vstinner I have ran into the same problem as @hashbrowncipher: some code (like the FTP-tests) do not expect the `SSLSocket.read()` to consume the EOF.
Do you agree with @hashbrowncipher that there should be a new parameter (for example on the SSLContext) to enable this new behaviour?
If you agree, then I will try to implement and also add the same for `SSLSocket.write()`.

**Comment by dimaqq** [comment]
my 2c, as an end-user:
I think the EOF handling is a leftover from bygone days;
Today the user code would rather open a new connection, thus `.read()` consuming EOF is totally reasonable...
Perhaps if that breaks code that calls `.read()` again after, it needs to re-report EOF or raise or something 🤔

**Pull request closed without merging**
